src/tensor_operations/matrix_decomposition.jl

```diff
@@ -360,44 +360,92 @@ function noinds_error_message(decomp::String)
    treating the ITensor as a matrix from the primed to the unprimed indices."
 end
 
+function add_trivial_index(A::ITensor, Ainds)
+  α = trivial_index(Ainds) #If Ainds[1] has no QNs makes Index(1), otherwise Index(QN()=>1)
+  vα = onehot(eltype(A), α => 1)
+  A *= vα
+  return A, vα, [α]
+end
+
+function add_trivial_index(A::ITensor, Linds, Rinds)
+  vαl, vαr = nothing, nothing
+  if isempty(Linds)
+    A, vαl, Linds = add_trivial_index(A, Rinds)
+  end
+  if isempty(Rinds)
+    A, vαr, Rinds = add_trivial_index(A, Linds)
+  end
+  return A, vαl, vαr, Linds, Rinds
+end
+
+function remove_trivial_index(Q::ITensor, R::ITensor, vαl, vαr)
+  if !isnothing(vαl) #should have only dummy + qr,Link
+    Q *= dag(vαl)
+  end
+  if !isnothing(vαr) #should have only dummy + qr,Link
+    R *= dag(vαr)
+  end
+  return Q, R
+end
+
 qr(A::ITensor; kwargs...) = error(noinds_error_message("qr"))
 
-# TODO: write this in terms of combiners and then
-# call qr on the order-2 tensors directly
 function qr(A::ITensor, Linds...; kwargs...)
-  tags::TagSet = get(kwargs, :tags, "Link,qr")
+  qtag::TagSet = get(kwargs, :tags, "Link,qr") #tag for new index between Q and R
   Lis = commoninds(A, indices(Linds...))
   Ris = uniqueinds(A, Lis)
-
-  Lis_original = Lis
-  Ris_original = Ris
-  if isempty(Lis_original)
-    α = trivial_index(Ris)
-    vLα = onehot(eltype(A), α => 1)
-    A *= vLα
-    Lis = [α]
-  end
-  if isempty(Ris_original)
-    α = trivial_index(Lis)
-    vRα = onehot(eltype(A), α => 1)
-    A *= vRα
-    Ris = [α]
+  lre = isempty(Lis) || isempty(Ris)
+  # make a dummy index with dim=1 and incorporate into A so the Lis & Ris can never
+  # be empty.  A essentially becomes 1D after collection.
+  if (lre)
+    A, vαl, vαr, Lis, Ris = add_trivial_index(A, Lis, Ris)
```

Comment:
Can we add the check for `isempty(Lis) || isempty(Ris)` inside `add_trivial_index` to simplify this a bit?